### PR TITLE
Added UDP client example application.

### PR DESCRIPTION
**Pull request by jimenezjose**
# New Features
- Added a C++ UDP client example

### CODE REVIEW

**Comment by jimenezjose** [comment]
Hey @adamshapiro0 is there a reason I am have 6 tests with status expected?

**Comment by adamshapiro0** [comment]
Yep sorry about that, it was waiting on me. It needs approval to run the tests for first time contributors, but apparently Github doesn't actually send any sort of notification when it is waiting.

FYI, we use a `git rebase` flow to make sure that all commits and merges are 100% linear for testing purposes (you'll see what I mean if you look at the git history with a GUI or with `git log --all --graph --decorate`). Would you mind please rebasing your branch on top of the latest `master` and removing the merge commit that's currently in there?

Once it's rebased let me know and I can look it over.

**Comment by jimenezjose** [comment]
@adamshapiro0 this `git rebase` process has not been straight forward. Can I get assistance one squashing all these commit messages onto just one? 

`git checkout udp-cpp-example`

Here is my attempt:
`git log --all --graph --decorate`
I want to combine commits:

* commit 158f776aa7f36da08b1126f798cd5f9b866db9d2 (origin/udp-cpp-example)
    `raw message output typo fix for debug mode`
*  commit 4289a2a2f2d82cae33b35e18005bb9bfe279abf5
    `Merge branch 'PointOneNav:master' into udp-cpp-example`
* commit 1448ec1b88c7fa935d1c17a60e0ff92571b4d34d
    `c++ udp example all setup`

I then did:
`git log --oneline`
To find the commit before `c++ udp example all setup`.

Result: 
1448ec1 `c++ udp example all setup`
5432d73 `Enabled style checking and Python unit tests in CI, and added contributor instructions.`

I then did:
`git rebase -i  5432d73`

such that:
pick e545e9b `Fixed UNKNOWN satellite type value in Python code.`
pick 1448ec1 `c++ udp example all setup`
squash 158f776 `raw message output typo fix for debug mode`

I couldn't find the merge commit but this is what I did to squash 158f776 to 1448ec1.

I then:
`git push --force origin udp-cpp-example`


Should I have removed your commits from being picked? Not sure what to do after this?

**Comment by adamshapiro0** [comment]
Sorry about that, it looks like the rebase may have gotten confused since you already had an unexpected merge commit in the middle of your branch. Our instructions could also use a bit of sprucing up for clarity.

There are really two totally separate operations here:
1. Rebasing your branch to update it with the latest upstream changes
2. Doing an _interactive_ rebase to squash your 2 commits together

The steps you described are correct for doing (2), but I believe the merge commit in between them may have thrown a wrench in the works. Normally that merge wouldn't be there and what you did would work just fine. You can read the command you wrote (`git rebase -i 5432d73`) as "do an interactive rebase and modify all commits starting _after_ commit 5432d73".

In order to update to the latest changes, you want to do (1). For that, you normally want to do `git rebase PointOneNav/master` (I used `PointOneNav/` here since your local repo is cloned from your fork so `origin/` is your fork on github). You'd read that as "rebase all commits on my current branch on top of the current version of `PointOneNav/master`".

Basically:
```
git checkout udp-cpp-example    # Checkout your branch
git fetch PointOneNav           # Fetch the latest changes from the upstream repo
git rebase PointOneNav/master   # Rebase your branch onto the latest version of the upstream repo
```

There are a few easy ways you can correct your branch. Since you already squashed stuff and now just have the one commit (currently 247c428), I'd suggest simply resetting your branch to sit at `PointOneNav/master` and then cherry picking your one commit:

```
git checkout udp-cpp-example
git fetch PointOneNav
git reset --hard PointOneNav/master
git cherry-pick 247c428
git push --force
```

**Comment by adamshapiro0** [comment]
@jimenezjose if you have a few minutes, would you mind taking a look at https://github.com/PointOneNav/fusion-engine-client/pull/72 (rendered version here: https://github.com/PointOneNav/fusion-engine-client/blob/instructions/CONTRIBUTING.md#updating-to-the-latest-changes-rebasing) to see if it's a little more clear? I appreciate any feedback. Thanks!

**Comment by jimenezjose** [comment]
@adamshapiro0 
I will documenting my entire process of success:

`git fetch PointOneNav` did not work. 

I resolved this with executing: 
`git remote add PointOneNav https://github.com/PointOneNav/fusion-engine-client.git`
before referencing 'PointOneNav'.

While rebasing to master there was a lot of manual work. Thankfully VSCode helped with this process.

Actually, it's a lot of manual work. Is there a way to automate the rebasing of 91 commits without breaking anything?

**Comment by adamshapiro0** [comment]
@jimenezjose that's helpful, I appreciate your patience and appreciate you documenting it :)

You should definitely not be rebasing 91 commits. You only have 1 commit. As I mentioned, since you already merged into your branch and then rebased, the branch is in a slightly strange state and the rebase may be getting confused. If you notice when it's getting stuck, it's actually now trying to recreate all of _my_ commits old, which already exist on `master`. That's definitely not what you wanted to do.

In this case, I would recommend _not_ rebasing, and instead doing the cherry pick method I suggested:
```
git checkout udp-cpp-example
git fetch PointOneNav
git reset --hard PointOneNav/master
git cherry-pick 247c428              <-- or whatever the correct commit hash is for commit: "c++ udp example all setup"
git push --force
```

Or, another way you could do this if you did want to use rebase is to first reset your branch to erase the merge commit that's currently on top of it, and then use `git rebase --onto` to specifically rebase just your one commit:
```
git checkout udp-cpp-example
git reset --hard 247c428              <-- or whatever the correct commit hash is for commit: "c++ udp example all setup"
git fetch PointOneNav
git rebase --onto PointOneNav/master HEAD~1
```

You can read that as "rebase my branch on top of PointOneNav/master, starting _after_ the commit just before mine (HEAD~1)".

**Comment by jimenezjose** [comment]
I will documenting my entire process of success: (continued)

```
git checkout udp-cpp-example
git remote add PointOneNav https://github.com/PointOneNav/fusion-engine-client.git
git fetch PointOneNav
git reset --hard PointOneNav/master
git log --all --graph --decorate                      # used to find/confirm the commit hash for commit: "c++ udp example all setup"
git cherry-pick 247c428
git push --force
```

**Comment by jimenezjose** [comment]
@adamshapiro0

File: examples/udp_client/linux_udp_client.cc [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+#include <cstring> // For memcpy()
+#include <string> // For stoi() and strerror()
+
+#include <netdb.h> // For gethostbyname() and hostent
+#include <netinet/in.h> // For IPPROTO_* macros and htons()
+#include <sys/socket.h> // For socket support.
+#include <unistd.h> // For close()
+
+#include <iostream>
+#include <fstream>
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:26]: Sort along with the other includes above please. Standard includes (iostream, fstream) in the first group, system includes (inet.h) in teh second.

File: examples/udp_client/linux_udp_client.cc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+/******************************************************************************/
+void HandleSignal(int signal) {
+  if (signal == SIGINT || signal == SIGTERM) {
+    std::signal(signal, SIG_DFL);
+    shutdown_pending_ = true;
+  }
+}
+
+void * get_in_addr(struct sockaddr * sa) {
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:48]: Formatting of *s not quite right here and below - please run `clang-format` on the file to get all the style stuff.

Comment (adamshapiro0) [reply examples/udp_client/linux_udp_client.cc:48]: Add a `/***/` bar before this function, similar to the other functions.

Function names should be title case (`GetInAddr`).

File: examples/udp_client/linux_udp_client.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    if (sa->sa_family == AF_INET) {
+        return &(((struct sockaddr_in *)sa)->sin_addr);
+    }
+    return &(((struct sockaddr_in6 *)sa)->sin6_addr);
+}
+/******************************************************************************/
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:54]: This bar belongs to `main()`, so please swap this line and the blank line just below.

File: examples/udp_client/linux_udp_client.cc [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+int main(int argc, const char* argv[]) {
+  // Parse arguments.
+  if (argc > 2) {
+    printf(R"EOF(
+Usage: %s [PORT]
+
+Connect to an Atlas device over UDP and print out the incoming message
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:62]: Unlike TCP, which connects to Atlas as a client, with UDP the data is sent by Atlas to the client. For clarity, I'd suggest rephrasing this more like "Receive incoming data from an Atlas device over UDP and..."

File: examples/udp_client/linux_udp_client.cc [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+contents.
+)EOF", 
+        argv[0]);
+    return 0;
+  }
+
+  int port = (argc == 2) ? std::stoi(argv[1]) : 12345;
+
+  // create UDP socket.
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:71]: Capitalize all comments please :)

File: examples/udp_client/linux_udp_client.cc [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  int sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (sock < 0) {
+    printf("Error creating socket.\n");
+    return 2;
+  }
+
+  // bind socket to port
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:78]: Same here: capital "Bind" and add a period. We like to make everything complete sentences wherever possible for consistency and readability.

File: examples/udp_client/linux_udp_client.cc [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+#include <iostream>
+#include <fstream>
+#include <arpa/inet.h>
+
+#include <point_one/fusion_engine/parsers/fusion_engine_framer.h>
+
+#include "../common/print_message.h"
+
+#define DEBUG_ON 0
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:33]: This doesn't seem to be used anymore - remove?

File: examples/udp_client/linux_udp_client.cc [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      ret = 4;
+      break;
+    } else if (bytes_read == 0) {
+      printf("Socket closed remotely.\n");
+      break;
+    }
+
+
+    inet_ntop(their_addr.ss_family, get_in_addr((struct sockaddr *)&their_addr), their_ip, sizeof(their_ip));
+    buffer[bytes_read] = '\0';
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:120]: This incoming data is almost always going to be binary, not a string, so printing it with `%s` below may do unexpected things and probably won't do anything too useful. I imagine this was here from your initial debugging? I'd suggest removing these lines and the if statement/print below now that this is all working. You can also then remove `their_addr` and `their_ip`, and use `recv()` instead of `recvfrom()`.

**Review by adamshapiro0 — CHANGES_REQUESTED** [review]
Nice work and much appreciated! Just some minor cleanup to take care of and this should be ready for merge.

Feel free to do the cleanup either as a separate commit, or by amending the exiting one if you prefer. Let me know once it's all set and pushed up.

**Comment by jimenezjose** [comment]
After resolving the code style I did the following:

```
git checkout udp-cpp-example
git remote add PointOneNav https://github.com/PointOneNav/fusion-engine-client.git
git fetch PointOneNav
git reset --hard PointOneNav/master
git pull
```

*my code style update*

```
git add .
git commit --amend
git push --force origin udp-cpp-example
```

File: examples/udp_client/linux_udp_client.cc (hunk 8 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:120]: As you mentioned before UDP, in contrast to TCP, is connectionless. Usually with TCP you have a connect() call to do the TCP 3-way handshake with the client and server but in this case the server is just spitting out packets with the ip address destined to the client machine. The client machine here just receives the packet on a specified port. That is really the only difference between TCP and UDP. In other words, `recvfrom()` is needed for UDP.

File: examples/udp_client/linux_udp_client.cc (hunk 7 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:32]: I think this debug statement should stay. Only because as the user of Atlas this helped me debug issues (usually on the server side) about the network topology between the Atlas and the listening linux client. Of course you are the owner of this codebase so it is your decision. But I'd say if you want to make the user's life easier... leave it.

Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:32]: Also, just thinking out load.. this also shows the developer that it is possible to receive packets from different ip addresses on the same port.

File: examples/udp_client/linux_udp_client.cc (hunk 1 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:25]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 2 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 3 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:47]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 4 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 5 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 6 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc]: done.

File: examples/udp_client/linux_udp_client.cc [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+  uint8_t buffer[1024];
+  size_t total_bytes_read = 0;
+  struct sockaddr_storage their_addr; // address of recieved packet
+  socklen_t addr_len = sizeof(their_addr);
+  ret = 0;
+  char their_ip[INET6_ADDRSTRLEN];
+  while (!shutdown_pending_) {
+    ssize_t bytes_read = recvfrom(sock, buffer, sizeof(buffer), 0,
+        (struct sockaddr *)&their_addr, &addr_len);
```
[the remark this answers is not in the document]
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 7 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:32]: Oh and also, the compiler would also complain about unused variables (e.g. `their_ip`).

File: examples/udp_client/linux_udp_client.cc (hunk 8 above)
Comment (adamshapiro0) [reply examples/udp_client/linux_udp_client.cc:120]: Technically, I believe you can use either `recv()` or `recvfrom()` for UDP if you are not interested in the sender's IP address (`recvfrom()` accepts `nullptr` for the source arguments). From the `recv(2)` man page:

> Also, the following call
>   recv(sockfd, buf, len, flags);
> is equivalent to
>   recvfrom(sockfd, buf, len, flags, NULL, NULL);

In this case, it's probably safe to assume that the packet is from the Atlas device since for UDP you must first configure Atlas to send to your machine specifically, so we don't really need the `their_*` information. That being said, if you think it's useful to keep the debug print then it's probably ok.

In that case though, it's definitely wrap the call in a `#if DEBUG_ON` instead of `if (DEBUG_ON)` -- since `DEBUG_ON` is a constant, using an if statement will produce a `Statement always evaluates to false` warning on some compilers.

File: examples/udp_client/linux_udp_client.cc [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+      printf("Socket closed remotely.\n");
+      break;
+    }
+
+    inet_ntop(their_addr.ss_family, GetInAddr((struct sockaddr*)&their_addr),
+              their_ip, sizeof(their_ip));
+    buffer[bytes_read] = '\0';
+    if (DEBUG_ON) {
+      // Enable to debug udp server to client problems.
+      printf("listener: received raw packet [%s] from %s\n", buffer, their_ip);
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:123]: If we are keeping this print, it is very important _not_ to print out `buffer` as a string. It's ok to print `bytes_read` perhaps, but `buffer` contains arbitrary binary data and interpreting it as a string can have unintended consequences.

File: examples/udp_client/linux_udp_client.cc (hunk 7 above)
Comment (adamshapiro0) [reply examples/udp_client/linux_udp_client.cc:32]: If you found it helpful, I'm ok with leaving it. Just see the notes below about using `#if` instead, and about not printing out `buffer`.

On the unused variables note, technically `recvfrom()` accepts `nullptr` for the `srcaddr` and `addrlen` arguments, so if you are not using them you don't need to specify them, or you can use `recv()` instead. See the note in https://man7.org/linux/man-pages/man2/recv.2.html.

File: examples/udp_client/linux_udp_client.cc (hunk 8 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:120]: I think that may just work. I just remember that I wanted to do just that in Vegas but for some reason I was having issues using recv() instead of recvfrom(). I would have used recv() but it was not working at the time. And I just remember the TCP client working with those system calls and noticing that the only difference was the connect() instead of bind() for UDP. After reading a little more I think it would be possible to use connect() for UDP transactions and then use recv() and send() bc the source and destination information will automatically be stored. Hmm I'd have to test it though. I wouldn't want to change something that I know works for a code style optimization that has not been tested. Ref: https://beej.us/guide/bgnet/html/index-wide.html#datagram, https://beej.us/guide/bgnet/html/index-wide.html#sendtorecv.

Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:120]: great point about the #if preprocessor condition.

File: examples/udp_client/linux_udp_client.cc (hunk 10 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 7 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:32]: done.

File: examples/udp_client/linux_udp_client.cc [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      printf("Socket closed remotely.\n");
+      break;
+    }
+
+    inet_ntop(their_addr.ss_family, GetInAddr((struct sockaddr*)&their_addr),
+              their_ip, sizeof(their_ip));
+  
+    #if DEBUG_ON
+      // Enable to debug udp server to client problems.
+      printf("listener: received raw packet [%d bytes] from %s\n", bytes_read, their_ip);
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:123]: `bytes_read` is a `size_t` so you should print it with `%zu` instead of `%d`, which is for (signed) `int`.

File: examples/udp_client/linux_udp_client.cc [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+      break;
+    } else if (bytes_read == 0) {
+      printf("Socket closed remotely.\n");
+      break;
+    }
+
+    inet_ntop(their_addr.ss_family, GetInAddr((struct sockaddr*)&their_addr),
+              their_ip, sizeof(their_ip));
+  
+    #if DEBUG_ON
```
Comment (adamshapiro0) [inline examples/udp_client/linux_udp_client.cc:121]: You may want rerun `clang-format` on the file. Normally it aligns preprocessor directives at the start of the line like
```
#if DEBUG_ON // yes

     #if DEBUG_ON // no
```

File: examples/udp_client/linux_udp_client.cc (hunk 11 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:123]: done.

File: examples/udp_client/linux_udp_client.cc (hunk 12 above)
Comment (jimenezjose) [reply examples/udp_client/linux_udp_client.cc:121]: done.
